sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -0,0 +1,429 @@
+        }
+
+        private static ResourceIdentifier Create(string resourceId)
+        {
+            Argument.AssertNotNullOrEmpty(resourceId, nameof(resourceId));
+
+            if (resourceId[0] != Separator)
+                throw new ArgumentOutOfRangeException(nameof(resourceId), "The ResourceIdentifier must start with '/'.");
+
+            if (resourceId.Length == 1)
+                return Root;
+
+            string[] parts = resourceId.Split(Separator);
+            if (parts.Length < 3)
+                throw new ArgumentOutOfRangeException(nameof(resourceId), "The ResourceIdentifier is too short it must have at least 2 path segments.");
+
+            if (!parts[1].Equals(SubscriptionsKey, StringComparison.OrdinalIgnoreCase) && !parts[1].Equals(ProvidersKey, StringComparison.OrdinalIgnoreCase))
+                throw new ArgumentOutOfRangeException(nameof(resourceId), $"The ResourceIdentifier must start with either '/{SubscriptionsKey}' or '/{ProvidersKey}'.");
+
+            int validLength = string.IsNullOrEmpty(parts[parts.Length - 1]) ? parts.Length - 1 : parts.Length;
+            return AppendNext(Root, parts, 1, validLength);
+        }
+
+        private static ResourceIdentifier AppendNext(ResourceIdentifier parent, string[] parts, int startIndex, int length)
+        {
+            int partsCount = length - startIndex;
+
+            if (partsCount == 0)
+                return parent;
+
+            string firstElement = parts[startIndex];
+
+            if (partsCount == 1)
+            {
+                //subscriptions and resourceGroups aren't valid ids without their name
+                if (firstElement.Equals(SubscriptionsKey, StringComparison.OrdinalIgnoreCase) || firstElement.Equals(ResourceGroupsLowerKey, StringComparison.OrdinalIgnoreCase))
+                    throw new ArgumentOutOfRangeException(nameof(parts), $"The ResourceIdentifier is missing the key for {firstElement}.");
+
+                //resourceGroup must contain either child or provider resource type
+                if (parent.ResourceType == ResourceGroupResourceType)
+                    throw new ArgumentOutOfRangeException(nameof(parts), $"Expected {ProvidersKey} path segment after {ResourceGroupsLowerKey}.");
+
+                return new ResourceIdentifier(parent, firstElement, string.Empty);
+            }
+
+            if (firstElement.Equals(ProvidersKey, StringComparison.OrdinalIgnoreCase))
+            {
+                if (partsCount == 2 || parts[startIndex + 2].Equals(ProvidersKey, StringComparison.OrdinalIgnoreCase))
+                {
+                    //provider resource can only be on a tenant or a subscription parent
+                    if (parent.ResourceType == SubscriptionResourceType || parent.ResourceType == TenantResourceType)
+                        return AppendNext(new ResourceIdentifier(parent, ProviderResourceType, parts[startIndex + 1], true), parts, startIndex + 2, length);
```

Comment:
From a maintainability standpoint, it might be nice to new the new parent here on a separate line.